src/algebraic_geometry/prime_spectrum.lean

```diff
@@ -0,0 +1,202 @@
+import ring_theory.ideal_operations
+
+/-!
+# Prime spectrum of a commutative ring
+
+The prime spectrum of a commutative ring is the type of all prime ideals.
+It is naturally endowed with a topology: the Zariski topology.
+
+(It is also naturally endowed with a sheaf of rings,
+but that sheaf is not constructed in this file.
+It should be contributed to mathlib in future work.)
+
+## Inspiration/contributors
+
+The contents of this file draw inspiration from
+https://github.com/ramonfmir/lean-scheme
+which has contributions from Ramon Fernandez Mir, Kevin Buzzard, Kenny Lau,
+and Chris Hughes (on an earlier repository).
+
+-/
+
+universe variables u v
+
+variables (R : Type u) [comm_ring R]
+
+/-- The prime spectrum of a commutative ring `R`
+is the type of all prime ideal of `R`.
+
+It is naturally endowed with a topology (the Zariski topology),
+and a sheaf of commutative rings (not yet in mathlib).
+It is a fundamental building block in algebraic geometry. -/
+def prime_spectrum := {I : ideal R // I.is_prime}
+
+variable {R}
+
+namespace prime_spectrum
+
+/-- A method to view a point in the prime spectrum of a commutative ring
+as an ideal of that ring. -/
+abbreviation as_ideal (x : prime_spectrum R) : ideal R := x.val
+
+instance as_ideal.is_prime (x : prime_spectrum R) :
+  x.as_ideal.is_prime := x.2
+
+@[ext] lemma ext {x y : prime_spectrum R} :
+  x = y ↔ x.as_ideal = y.as_ideal :=
+subtype.ext
+
+/-- The zero locus of a set `s` of elements of a commutative ring `R`
+is the set of all prime ideals of the ring that contain the set `s`.
+
+An element `f` of `R` can be thought of as a dependent function
+on the prime spectrum of `R`.
+At a point `x` (a prime ideal)
+the function (i.e., element) `f` takes values in the quotient ring `R` modulo the prime ideal `x`.
+In this manner, `zero_locus s` is exactly the subset of `prime_spectrum R`
+where all "functions" in `s` vanish simultaneously.
+-/
+def zero_locus (s : set R) : set (prime_spectrum R) :=
+{x | s ⊆ x.as_ideal}
+
+@[simp] lemma mem_zero_locus (x : prime_spectrum R) (s : set R) :
+  x ∈ zero_locus s ↔ s ⊆ x.as_ideal := iff.rfl
+
+lemma zero_locus_empty_of_one_mem (s : set R) (h : (1:R) ∈ s) :
+  zero_locus s = ∅ :=
+begin
+  rw set.eq_empty_iff_forall_not_mem,
+  intros x hx,
+  rw mem_zero_locus at hx,
+  have x_prime : x.as_ideal.is_prime := by apply_instance,
+  have eq_top : x.as_ideal = ⊤, { rw ideal.eq_top_iff_one, exact hx h },
+  apply x_prime.1 eq_top,
+end
+
+@[simp] lemma zero_locus_univ :
+  zero_locus (set.univ : set R) = ∅ :=
+zero_locus_empty_of_one_mem _ (set.mem_univ 1)
+
+@[simp] lemma zero_locus_span (s : set R) :
+  zero_locus (ideal.span s : set R) = zero_locus s :=
+begin
+  ext x,
+  simp only [mem_zero_locus],
+  split,
+  { exact set.subset.trans ideal.subset_span },
+  { intro h, rwa ← ideal.span_le at h }
+end
+
+lemma union_zero_locus_ideal (I J : ideal R) :
+  zero_locus (I : set R) ∪ zero_locus J = zero_locus (I ⊓ J : ideal R) :=
```

Comment:
Do we have a name for this property on lattices?

Reply:
I guess this is a consequence of some galois connection (contravariant).